### PR TITLE
Make CMAC properly fail if cipher is not CBC mode one

### PR DESCRIPTION
**Pull request by t8m**
Also avoid misleading users in the demo by use of the alias name.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
I think it is a bug that we allow non CBC mode ciphers with CMAC as that is completely insecure use of the crypto primitive. So I propose this should be also added to 3.0.

**Review by beldmit — APPROVED** [review]
LGTM if tests passed

**Comment by t8m** [comment]
Heh, we misuse CMAC in evp_extra_test.c. Fixup commit added.

**Comment by t8m** [comment]
@beldmit still OK?

**Comment by beldmit** [comment]
Yes, still OK if tests passed

**Comment by mattcaswell** [comment]
> we allow non-CBC-mode ciphers with CMAC

Is this in the RFC? So CFB/OFB modes, for example, are not ok? I'm wondering how breaking this change might end up being if we backport to 3.0...

**Comment by mattcaswell** [comment]
CI failure looks relevant

**Comment by t8m** [comment]
> Is this in the RFC? So CFB/OFB modes, for example, are not ok? I'm wondering how breaking this change might end up being if we backport to 3.0...

This is by definition of the CMAC.  https://en.wikipedia.org/wiki/One-key_MAC
And yes, it is also in the AES-CMAC definition in RFC 4493.

**Comment by t8m** [comment]
> Is this in the RFC? So CFB/OFB modes, for example, are not ok? I'm wondering how breaking this change might end up being if we backport to 3.0...

In theory one could use the CMAC MAC with other mode cipher but the result would be interoperable and with completely bogus security properties - potentially even completely insecure one.

**Comment by mattcaswell** [comment]
> In theory one could use the CMAC MAC with other mode cipher but the result would not be interoperable and with completely bogus security properties - potentially even completely insecure one

That may be the case - but my bet is that this will break people. Not a decision to make without some OTC input IMO.

**Comment by mattcaswell** [comment]
** OTC Question: Should we fix this, and if so in which branches? **

File: providers/implementations/macs/cmac_prov.c
Comment (paulidale) [inline providers/implementations/macs/cmac_prov.c:204]: Nicely backwards compatible if the returned error is ignored :smiley:

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
**OTC: This is ok for 3.1 and above, not for 3.0.**

**Comment by t8m** [comment]
Actually this needs re-approval from @beldmit

**Comment by beldmit** [comment]
Still OK

**Comment by t8m** [comment]
Merged to master and 3.1 branches. Thank you for the reviews.

**Pull request closed without merging**
